platforms/ios/framework/src/TGMapView.mm

```diff
@@ -1,44 +1,320 @@
+    _panGestureRecognizer.maximumNumberOfTouches = 1;
+
+    _pinchGestureRecognizer = [[UIPinchGestureRecognizer alloc] initWithTarget:self action:@selector(respondToPinchGesture:)];
+    _rotationGestureRecognizer = [[UIRotationGestureRecognizer alloc] initWithTarget:self action:@selector(respondToRotationGesture:)];
+    _shoveGestureRecognizer = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(respondToShoveGesture:)];
+    _shoveGestureRecognizer.minimumNumberOfTouches = 2;
+    _longPressGestureRecognizer = [[UILongPressGestureRecognizer alloc] initWithTarget:self action:@selector(respondToLongPressGesture:)];
+
+    // Use the delegate method 'shouldRecognizeSimultaneouslyWithGestureRecognizer' for gestures that can be concurrent
+    _panGestureRecognizer.delegate = self;
+    _pinchGestureRecognizer.delegate = self;
+    _rotationGestureRecognizer.delegate = self;
+
+    [_glView addGestureRecognizer:_tapGestureRecognizer];
+    [_glView addGestureRecognizer:_doubleTapGestureRecognizer];
+    [_glView addGestureRecognizer:_panGestureRecognizer];
+    [_glView addGestureRecognizer:_pinchGestureRecognizer];
+    [_glView addGestureRecognizer:_rotationGestureRecognizer];
+    [_glView addGestureRecognizer:_shoveGestureRecognizer];
+    [_glView addGestureRecognizer:_longPressGestureRecognizer];
+}
+
+- (void)displayLinkUpdate:(CADisplayLink *)sender
+{
+    if (_renderRequested || self.continuous) {
+        _renderRequested = NO;
+        [self.glView display];
+    }
+}
+
+
+- (void)didReceiveMemoryWarning
+{
+    if (self.map) {
+        self.map->onMemoryWarning();
+    }
+}
+
+- (void)didMoveToWindow
+{
+    [self setupDisplayLink];
+    [super didMoveToWindow];
+}
+
+- (void)didMoveToSuperview
+{
+    [self setupDisplayLink];
+    [super didMoveToSuperview];
+}
+
+- (void)layoutSubviews
+{
+    [super layoutSubviews];
+
+    CGSize size = self.bounds.size;
+    self.map->resize(size.width * self.contentScaleFactor, size.height * self.contentScaleFactor);
+
+    [self requestRender];
+}
+
+- (void)requestRender
+{
+    if (!self.map) { return; }
+
+    _renderRequested = YES;
+}
+
+- (void)captureScreenshot:(BOOL)waitForViewComplete
+{
+    self->_captureFrameWaitForViewComplete = waitForViewComplete;
+    self->_shouldCaptureFrame = YES;
+}
+
+- (void)glkView:(GLKView *)view drawInRect:(CGRect)rect
+{
+    if (self->_viewInBackground) {
+        return;
+    }
+
+    CFTimeInterval dt = _displayLink.targetTimestamp - _displayLink.timestamp;
+    _viewComplete = self.map->update(dt);
+
+    self.map->render();
+
+    if (_viewComplete && [self.mapViewDelegate respondsToSelector:@selector(mapViewDidCompleteLoading:)]) {
+        [self.mapViewDelegate mapViewDidCompleteLoading:self];
+    }
+
+    if (self.mapViewDelegate && [self.mapViewDelegate respondsToSelector:@selector(mapView:didCaptureScreenshot:)]) {
```

Comment:
@msmollin In the conditional here, is it necessary to check the delegate for nullity first? I've read that you can just send `respondsToSelector` to a `nil` receiver and you'll just get `NO`, but I don't know whether this is a common/good practice.

Reply:
Yeah that's fine. The nil check first is unnecessary.